### ISSUE TITLE
Translation, Gym helper

### ISSUE THREAD

**Issue by Dreckiger-Dan** [farkam135/GoIV#69]
I am currently working on putting all strings into the `strings.xml`.

What do you think about showing a list of pokémon that are good against the pokémon in the gym on the gym screen? I'm going to try to implement it once I have extracted all the string.

**Comment by nahojjjen** [comment]
Could you elaborate what you mean? Do you mean type advantages?

A sketch of how the feature could work (paint sketch is fine) to help us understand would be great.

**Comment by Dreckiger-Dan** [comment]
![gym](https://cloud.githubusercontent.com/assets/10210133/17767409/2cac3556-6530-11e6-8677-a47eab1a86d9.jpg)

**Comment by Dreckiger-Dan** [comment]
I think, I'm going to use the white outline of the gym picture and the platform the trainer is standing on to check if the screenshot is a gym-screen...

@farkam135 how did you calculate the [`areaX1`, `areaY1`, ...](https://github.com/farkam135/GoIV/blame/master/app/src/main/java/com/kamron/pogoiv/MainActivity.java#L220) values?

**Comment by sarav** [comment]
This feature would be very useful! I can't remember all the different type advantages and end up picking terrible ones. Btw, isn't Jolteon technically good against Vaporeon, but it's still a bad choice in PoGo? What data should this Weak/Strong against use?

A lot of people rename their pokemon though. I guess this is better than nothing.

**Comment by hlplulustg** [comment]
I wonder what information would be most useful. There's not only the aspects seen in the picture. It's more to consider:

- Attack (of gym pokemon) - Supereffective dmg x2
- Attack (of gym pokemon) - Weak dmg x0.5
- Defense (of gym pokemon) - Weakness dmg taken x2
- Defense (of gym pokemon) - Strong dmg x0.5
- Defense (of gym pokemon) - Resistant dmg taken x0
- two extra Defense combinations for STAB?

That's information on the defending pokemon (which is all the app could supply). Now the best attacking pokemon would be utilizing the defending pokemons Weakness for its moves and the defending pokemons Weak attack for the choice of type. Only if the combination is applicable of course.

The type advantage charts seem complimentary but unfortunaley they are not.

> What data should this Weak/Strong against use?
Based on that information people can choose a corresponding pokemon for attacking (utiling further informations such as moveset)

> A lot of people rename their pokemon though.
The names are only visible for the trainers themselves so this wouldn't be a problem.

**Comment by nahojjjen** [comment]
Just fyi
Type advantage isnt x2 dmg in pokemon go, its x1.25

https://www.reddit.com/r/TheSilphRoad/comments/4t8seh/pokemon_go_type_advantage_chart/

**Comment by hlplulustg** [comment]
Right, they lowered that for pokemon go and resistance is gone. Comment adjusted.

**Comment by sarav** [comment]
Here's another good resource for Pokemon data. It's well organized and well explained.

https://pokemongo.gamepress.gg/pokemon-list

**Comment by Dreckiger-Dan** [comment]
I'm not super familiar with android dev. Would it be a bad idea to create a new `Service` where all the screenshot and scanning code will be put or is it bad to create multiple services?

**Comment by sarav** [comment]
Anything that's not actively UI related seems like something that should be in a service or a pure helper class. Anything that's interacting with the user should be in an Activity. I'm not an android dev either, but that just feels right.

The stuff like scanning the bitmap, etc doesn't seem like it should be an activity or a service. Those just seem like helper classes. I think the Pokemon class should have constructors that could take the pokemon level, HP, etc and evaluates the various IVs possibilities and the rest of the details. An OCRhelper class to process bitmaps and return the values, etc.

Only the screen monitoring seems like something that should be in a service. The OCR result screen which allows the user to fix the values, etc seems like it should be a floating activity.

**Comment by farkam135** [comment]
@Dreckiger-Dan sorry for the super late response, i calculated those areas using my own device and then just finding the ratio in relation to width and height. Also i'd listen to @sarav, i'm not an android dev either and GoIV's codebase is really messy and needs a lot of outsourcing, I know it's a terrible way to do it but once all the base features are working then i'll go ahead and move things around, but I recommend with your feature just start off with good practices!

**Comment by sarav** [comment]
I'm kinda hijacking this issue since it's talking about refactoring. @nahojjjen, Thanks a lot for the refactoring. It's so much better and readable. I understood the code before, but didn't have the time to clean it up before making changes. Now it's a lot more manageable.

**Comment by nahojjjen** [comment]
I'm glad my efforts weren't wasted

**Comment by sarav** [comment]
I even made a commit that got merged. I think that was a fix for your refactor though :)

**Comment by sarav** [comment]
@nahojjjen You seem to be the OCR expert. Want to give this a shot next? This would be very useful for people like me (when I actually play the game instead of coding GoIV) who can't remember the weakness types for Pokemon.

**Comment by nahojjjen** [comment]
@sarav is what you're requesting that I make the screen scan detect when a user is viewing a gym?

**Comment by sarav** [comment]
Well, the whole feature :D 

I'm not a UI expert either. Obviously, it's your choice to work or not on this. But I think it's better to have you spend time on something only you can do.

**Comment by Dreckiger-Dan** [comment]
little update:
I currently have no time to work on the project because of school/work. I also don't play Pokemon Go that much anymore. Maybe it is better if somebody else implements this feature... Sorry for letting you wait for so long :pensive:

**Comment by rpgrca** [comment]
Someone at reddit said they wanted to create a GoIV-like framework where plugins could be added, I think that would be the best approach for this. One plugin would be to work with IV, another plugin would be the gym helper, another plugin could be potion-manager (optimize usage of potions across hurt pokemon, for example), pokemon capture location logging, etc.

**Comment by sarav** [comment]
Ehhh... I don't think the gym helper needs to be a plugin. That seems to fall under GoIV's main strength - detecting screens and doing OCR. But for the data analysis (off game and off app) and storage part, yeah, I agree that plugins might work.

**Comment by rpgrca** [comment]
My bad. I was thinking gym-helper was actually the data analysis of the information obtained, whereas GoIV would give a framework (Go! for example) from where it would trigger the different registered plugins (IV, gym-helper, etc).

GoIV is actually about analyzing IV, so I feel somewhat uncomfortable when I read about adding stuff that's not actually IV-related to the main application (and thus would prefer if a separation was made, for example via plugins). I have no idea where PokemonGo would go from now, but if they eventually add gender to pokemon and the ability to breed and trade (just like Pokemon Diamond/Pearl had), a branch of players might be only interested in the IV/moveset analysis to breed perfect pokemon, and might get bothered by having extra layers of processing (like gym helper, for example). So, it's important to let the user turn on/off each of them individually.

**Comment by sarav** [comment]
It won't be any extra layer. Just don't press the button on the gym view? Anyway, we are debating a hypothetical case. It's not like we are writing a plugin framework right now.
